src/main/java/life/qbic/omero/BasicOMEROClient.java

```diff
@@ -174,6 +181,117 @@ public void connect() {
     this.connect(this.username, this.password, this.hostname, this.port);
   }
 
+  /**
+   * This function imports an image file to an omero server.
+   * The omero dataset to which the image is registered to is fetched based on the openBIS identifiers.
+   *
+   * @param filePath the path to the image file
+   * @param projectId the project ID in OpenBIS
+   * @param sampleCode the sample Code in OpenBIS
+   * @return newly generated omero IDs for registered images
+   */
+  public Set<Long> registerImageFile(String filePath, String projectId, String sampleCode) {
+    Long datasetId = findOmeroDataset(projectId, sampleCode);
+    if (datasetId == null) {
+      return new HashSet<>();
+    }
+    return registerImageFile(filePath, datasetId);
+  }
+
+  /**
+   * This function imports an image file to an omero server.
+   * The id for the omero dataset to which the image is registered has to be provided.
+   *
+   * @param filePath the path to the image file
+   * @param datasetId the project ID in omero
+   * @return newly generated omero IDs for registered images
+   */
+  public Set<Long> registerImageFile(String filePath, long datasetId) {
+    ProcessBuilder builder = new ProcessBuilder("omero import",
+        "-s", hostname,
+        "-p " + port,
+        "-u", username,
+        "-w", password,
+        "-d " + (int) datasetId,
+        filePath
+    );
+
+    try {
+      Process importProcess = builder.start();
+      if (importProcess.exitValue() == 0) {
+        final String idLinePrefix = "Image:";
+        List<String> relevantOutput =
+            new BufferedReader(
+                  new InputStreamReader(importProcess.getInputStream())
+            ).lines().filter(line -> line.startsWith(idLinePrefix)).map(line -> line.substring(idLinePrefix.length())).collect(Collectors.toList());
+
+        if (relevantOutput.isEmpty()) {
+          return null;
+        }
+
+        Set<Long> imageIds = Arrays.stream(relevantOutput.get(0).split(",")).map(Long::parseLong).collect(
+            Collectors.toSet());
+        return imageIds;
+
+      } else {
+        // nothing was imported so we also return no IDs
+        return null;
+      }
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    return null;
+  }
+
+  /**
+   * Searches for an omero project for the current user
+   * @param projectName the name of the project which should be searched
+   * @param datasetName the name of the dataset that is searched
+   * @return the omero ID for the dataset matching search criteria, null if the search found no match
+   */
+  private Long findOmeroDataset(String projectName, String datasetName) {
+    if (!this.isConnected()) {
+      this.connect();
+    }
+    try {
+      BrowseFacility browseFacility = gateway.getFacility(BrowseFacility.class);
+      Collection<ProjectData> projects = browseFacility.getProjects(securityContext);
```

Comment:
has this been tested with different projects? I had a lot of trouble finding the correct stuff dependent on which user (or no user) I was using

Reply:
the user group information should be contained in the securityContext, thus no user is given explicitly. We want to always use the same user/password/hostname/port per client

Reply:
sounds good then